flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -1574,6 +1574,12 @@ public void run() {
                 // we need do continue despite that
                 try {
                     invokable.cancel();
+                    // Wait for any active actions to complete (e.g. timers, mailbox actions)
+                    // Before that, interrupt to notify them know about cancellation
+                    if (invokable.shouldInterruptOnCancel()) {
```

Comment:
I am wondering if it is possible that we end up hanging on the `terminationFuture`. Imagine a case where we do not finish cancelling before issuing the interruption signal. If I understand the code correctly there is no place where we handle the interruption signal and possibly finalize the termination future.

WDYT?

Reply:
Yes, you are right. If the interrupt is ignored and `taskCancellationTimeout` is NOT configured
then network resources will not be released.
I'm going to add a timeout to `join`. The default of 3 minutes for `taskCancellationTimeout` will work here too: in the absense of `taskCancellationTimeout` some delay in releasing resources is expected I think.

(note that invokables that spawn additional threads (like `SourceStreamTask`) their `cancelTask` should interrupt those threads)